include/runtime/header.h

```diff
@@ -336,19 +337,27 @@ void serialize_configurations(
 void serialize_configuration(
     block *subject, char const *sort, char **data_out, size_t *size_out,
     bool emit_size, bool use_intern);
-void serialize_configuration_v2(FILE *file, block *subject, uint32_t sort);
 void serialize_configuration_to_file(
     FILE *file, block *subject, bool emit_size, bool use_intern);
-void serialize_configuration_to_file_v2(FILE *file, block *subject);
-void write_uint64_to_file(FILE *file, uint64_t i);
-void write_bool_to_file(FILE *file, bool b);
 void serialize_term_to_file(
     FILE *file, void *subject, char const *sort, bool use_intern,
     bool k_item_inj = false);
-void serialize_term_to_file_v2(FILE *file, void *subject, uint64_t, bool);
 void serialize_raw_term_to_file(
     FILE *file, void *subject, char const *sort, bool use_intern);
-void print_variable_to_file(FILE *file, char const *varname);
+
+// The following functions are called by the generated code and runtime code to
+// ouput the proof trace data.
+void serialize_configuration_to_proof_trace(
+    kllvm::proof_trace_writer *proof_writer, block *subject, uint32_t sort);
+void serialize_configuration_to_proof_writer(
+    kllvm::proof_trace_writer *proof_writer, block *subject);
+void write_uint64_to_proof_trace(
+    kllvm::proof_trace_writer *proof_writer, uint64_t i);
+void write_bool_to_proof_trace(kllvm::proof_trace_writer *proof_writer, bool b);
+void write_string_to_proof_trace(
+    kllvm::proof_trace_writer *proof_writer, char const *str);
+void serialize_term_to_proof_trace(
+    kllvm::proof_trace_writer *proof_writer, void *subject, uint64_t, bool);
```

Comment:
These functions are marked as `extern "C"` but now take a pointer to a C++ type as an argument. This is fine and won't break anything, but if you were ever to move the declarations to a context where they'd actually be compiled as C, it would break things.

My suggestion would be to make the writer arguments `void *` and have an explicit static cast back to the writer type internally to the C++ implementation. This keeps the C declaration / C++ implementation boundary a bit cleaner.

Reply:
I have addressed this in my latest commit.

Reply:
Go ahead and merge in that case!

Reply:
I want Dwight to review this before I merge because it touched a lot of his code. But thank you for taking the time and giving feedback!